### PR TITLE
User and Group API tests

### PR DESCRIPTION
**Pull request by vijetmahabaleshwar-okta**
- Added tests for User and Group APIs
- Removed some of the redundant tests

### CODE REVIEW

**Comment by robertjd** [comment]
@vijetmahabaleshwar-okta since others will now be using this as reference, let's add these to this PR:

* Choose a "slug" for the scenario, the base URL string that will be used to identify this test to the mock server
* The ability to "turn on/turn off" those slugs, since we won't need them when we talk to the prod api (like we are now)

File: test/it/client-listUsers.js [hunk 1]
```diff
@@ -25,7 +30,7 @@ describe('client.listUsers()', () => {
   });
 
   it('should return a collection', () => {
-    assert.instanceOf(client.listUsers(), require('../../src/collection'));
+    expect(client.listUsers()).to.be.an.instanceof(require('../../src/collection'));
```
Comment (robertjd) [inline test/it/client-listUsers.js]: Let's move the require statements to the top of the file.

Comment (vijetmahabaleshwar-okta) [inline test/it/client-listUsers.js]: Done

File: test/it/user-with-password-update.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    expect(response.provider.type).to.equal('OKTA');
+
+    // 4. Update the user password through updated recovery question
+    userCredentials = {
+      password: { value: 'Abcd!234' },
+      recovery_question: { answer: 'forty two' }
+    };
+
+    await createdUser.forgotPassword(userCredentials, sendEmail);
+    response = await utils.authenticateUser(createdUser.profile.login, 'Abcd!234', client);
```
Comment (bdemers) [inline test/it/user-with-password-update.js]: We cannot support this as part of the _official_ ITs yet

**Comment by vijetmahabaleshwar-okta** [comment]
@bdemers - Regarding your comment on replacing `getFakeUser()`, I have submitted a patch that removes this. I have created a user for each scenario based on the scenario name. Please take a look at it and let me know if we can agree on the user and group names.
Regarding the cleanup after tests, we discussed about it in the context of mocha tests. Since we have to maintain the order of the calls, we decided that the cleanup would happen within the test itself rather than a `setup()` or `cleanup()` block. So, the user/group would be created in the first step witin each it() test block, and then it would be deleted in the last step of the same test block. And since each test has its own user/group, there is no dependency on other tests running. Hope that makes sense. 
But you can definitely use the setup()/cleanup() methods in your java tests.

**Comment by bdemers** [comment]
I was just stubbing out the 'user-with-recovery-question' test for the Java SDK I think this test is doing to many things.

I'm all for creating complex scenarios when needed, but I don' think this one can be broken down into a few different (more targeted) test cases.

For example this test could be broken up into the following:

Scenario: New user retrieved by Id
  When a new user is created
  Then the user can be retrieved by id

Scenario: New user found in search
  When a new user is created
  Then the user is found in search

**NOTE:** actual search filter is NOT important, as those scenarios are dependent on the server (not the client)

Scenario: User deleted
  Given a user exists
  When the user is deleted
  Then user is not found by search
    And user cannot be retrieved by id
    And user cannot be retrieved by email

Scenario: Update user password with recovery question.
  Given a user exists with security question
  When password is updated
  Then user can login successfully 
**NOTE:** we cannot implement this scenario currently

Scenario: Activate a user
  Given an inactive user exists
  When a user is activated
  Then user is marked as activated
**NOTE:** searching for the user is not important, as that is dependent on server functionality (nothing specific to the SDKs)

Also note, I'm not suggesting we use gherkin, just using it to illustrate the different scenarios.

**Comment by bdemers** [comment]
@vijetmahabaleshwar-okta I don't think the cleanup order location should be important to the tests.
When running against a live server the order is not important, and when running against a mock server it is not needed.

File: test/.eslintrc [hunk 3]
```diff
@@ -0,0 +1,87 @@
+{
+  "parserOptions": {
+    "ecmaVersion": 8 // or 2017
```
Comment (lboyette-okta) [inline test/.eslintrc]: I'm surprised eslint parses this as valid json. Maybe it treats it as yaml

File: test/it/client-listUsers.js [hunk 4]
```diff
@@ -1,11 +1,16 @@
-const assert = require('chai').assert;
-const faker = require('faker');
+const expect = require('chai').expect;
 const okta = require('../../');
 const models = require('../../src/models');
 const utils = require('../utils');
+const collection = require('../../src/collection');
+let orgUrl = process.env.OKTA_CLIENT_ORGURL;
```
Comment (lboyette-okta) [inline test/it/client-listUsers.js]: we should revisit this in the future. It'd be nice to use the same config parsing util that the client uses. If a customer wants to run the tests against an org that's setup via a config file, the env var solution won't work

File: test/it/client-listUsers.js [hunk 5]
```diff
@@ -15,8 +20,18 @@ describe('client.listUsers()', () => {
...
+        firstName: 'John',
+        lastName: 'List-Users',
+        email: 'john-list-users@example.com',
+        login: 'john-list-users@example.com'
+      },
+      credentials: {
+        password: {value: 'Abcd1234'}
+      }
+    };
+    newUser.profile.nickName = 'johny-list-users';
```
Comment (lboyette-okta) [inline test/it/client-listUsers.js]: nit: is there a reason we don't include the `nickName` in `profile` above? If it's to show that we're adding an unmapped property, let's add a comment. It looks a little weird without the explanation.

File: test/it/group-list.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  orgUrl = `${orgUrl}/list-groups`;
+}
+
+const client = new okta.Client({
+  orgUrl: orgUrl,
+  token: process.env.OKTA_CLIENT_TOKEN
+});
+
+describe('Group API tests', () => {
+  it('should list the groups and find the created group', async () => {
```
Comment (lboyette-okta) [inline test/it/group-list.js]: oh, cool! I didn't realize we could do async anonymous functions

File: test/it/client-listUsers.js [hunk 7]
```diff
@@ -97,16 +112,16 @@ describe('client.listUsers().each()', () => {
...
+      expect(err).to.equal('foo error');
     });
   });
 });
 
 describe('client.listUsers().next()', () => {
   it('should return User models', () => {
     return client.listUsers().next().then(result => {
-      assert.instanceOf(result.value, models.User);
+      expect(result.value).to.be.an.instanceof(models.User);
```
Comment (lboyette-okta) [inline test/it/client-listUsers.js]: in a follow-up PR, it'd be nice to convert all of these to async anonymous functions to match the other tests

File: test/it/group-update.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    // 2. Update the group name and description
+    const updateGroup = {
+      profile: {
+        name: 'Update Test Group - Updated',
+        description: 'Description updated',
+      }
+    };
+
+    const updatedGroup = await client.updateGroup(createdGroup.id, updateGroup);
```
Comment (lboyette-okta) [inline test/it/group-update.js]: This is interesting. We're not updating the `createdGroup` and resending it. That means our tape will not include the `id` property in the expected body. Some SDK's, like the PHP SDK, may not allow sending a model without an ID, so those SDKs couldn't pass this test on the mock-okta server.

We should modify `createdGroup` with the new `name` and `description`, then send that.

File: test/.eslintrc (hunk 3 above)
Comment (vijetmahabaleshwar-okta) [inline test/.eslintrc]: I could remove the commented part as its not needed anyway.

File: test/it/client-listUsers.js (hunk 5 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/client-listUsers.js]: @robertjd - This is one of your original tests. Do you have a reason for updating the profile `nickname`? I could add a comment anyway.

File: test/it/client-listUsers.js (hunk 7 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/client-listUsers.js]: Yes, I have that on my radar. This was Robert's original test where we didn't use async await feature. Will convert this in a later PR.

File: test/it/group-list.js (hunk 6 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/group-list.js]: async await is uber cool! 👍

File: test/it/group-update.js (hunk 8 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/group-update.js]: Done

**Comment by vijetmahabaleshwar-okta** [comment]
Code coverage is at ~80% with integration tests. Still working on `group rule` tests which should increase the coverage.
code-coverage

**Comment by vijetmahabaleshwar-okta** [comment]
Coverage increased to ~86% after adding `groupRule` test. :dancing_men: 
code-coverage

File: test/it/client-listUsers.js [hunk 9]
```diff
@@ -51,7 +67,7 @@ describe('client.listUsers().each()', () => {
   it('should return User models', () => {
     return client.listUsers().each(user => {
```
Comment (nbarbettini) [inline test/it/client-listUsers.js]: Question for @vijetmahabaleshwar-okta or @lboyette-okta - do these tests also make calls that would need to be handled by the test-server (returning taped responses)?

My implementation of async collections is different, and wouldn't require tests for each(), iteration abort, etc. If my tests **didn't** make these additional repeated calls to `/api/v1/users`, would the test-server think I didn't finish the scenario?

File: test/it/client-listUsers.js [hunk 10]
```diff
@@ -25,7 +41,7 @@ describe('client.listUsers()', () => {
   });
```
Comment (nbarbettini) [inline test/it/client-listUsers.js]: How will the test-server handle matching requests (like `DELETE /api/v1/users/{id}` where an arbitrary ID is included in the request?

Comment (vijetmahabaleshwar-okta) [inline test/it/client-listUsers.js]: In record mode, the ID for the user will be stored in the tape.
In playback mode, when you create a user, you will get back the same user ID that was used, when the user was created in playback mode. (ID would be returned by the real server during playback mode)
So, you wouldn't really include an arbitrary ID in the request. You'd use the ID that was returned to your by the response in a tape. 
@lboyette-okta - Does this sound right?

File: test/it/client-listUsers.js (hunk 9 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/client-listUsers.js]: Yes, I have tested this with taped responses. So, the tests do make calls that would be handled by test-server.
Regarding your second question, the test isn't really making repeated calls to `/api/v1/users`. It just makes 1 call, which returns a collection and then we iterate over it to ensure we have the right models. So, your implementation doesn't really matter as long as you make 1 call to `/api/v1/users`. Your test can then work on the collection return in any manner you deem right.

File: test/it/user-profile-update.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    utils.validateUser(createdUser, newUser);
+
+    // 2. Update the user profile
+    createdUser.profile.nickName = 'Batman';
+    const profileUpdateUser = await createdUser.update();
+    expect(profileUpdateUser.lastUpdated).to.be.gt(createdUser.lastUpdated);
+    expect(profileUpdateUser.profile.nickName).to.equal('Batman');
+
+    // 3. Delete the user
+    await utils.deleteUser(createdUser);
```
Comment (bdemers) [inline test/it/user-profile-update.js]: We should do the delete in a "cleanup" method (or some way we can not require it run at the end of the method)

*NOTE:* we will still need to do the cleanup, but this delete is not related to this specific test

Comment (bdemers) [inline test/it/user-profile-update.js]: Also cleanup needs a 'deactivate' call or this test will not be able to run a second time against a live server

File: test/it/user-profile-update.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      credentials: {
+        password: { value: 'Abcd1234' }
+      }
+    };
+
+    let queryParameters = { activate : 'false' };
+    const createdUser = await client.createUser(newUser, queryParameters);
+    utils.validateUser(createdUser, newUser);
+
+    // 2. Update the user profile
```
Comment (bdemers) [inline test/it/user-profile-update.js]: Need to wait 1 second here as that is the minimum time resolution of the 'lastUpdated' field

File: test/it/user-profile-update.js (hunk 11 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/user-profile-update.js]: utils.deleteUser() method is actually deactivating the user first and then deleting it. The method name isn't too intuitive I guess.

File: test/it/user-profile-update.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    createdUser.profile.nickName = 'Batman';
+    const profileUpdateUser = await createdUser.update();
+    expect(profileUpdateUser.lastUpdated).to.be.gt(createdUser.lastUpdated);
+    expect(profileUpdateUser.profile.nickName).to.equal('Batman');
```
Comment (bdemers) [inline test/it/user-profile-update.js]: This check is _less_ valid for the 'update' method if `update()` return itself.
for example in the Java SDK, this is equivalent to:

```
createdUser.profile.nickName = 'Batman'
assert createdUser.profile.nickName == 'Batman'
```

File: test/it/user-profile-update.js (hunk 12 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/user-profile-update.js]: Done

File: test/it/client-listUsers.js (hunk 10 above)
Comment (lboyette-okta) [inline test/it/client-listUsers.js]: That's correct!

File: test/it/user-profile-update.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+      await utils.cleanup(client, newUser);
+    }
+
+    let queryParameters = { activate : 'false' };
+    const createdUser = await client.createUser(newUser, queryParameters);
+    utils.validateUser(createdUser, newUser);
+
+    // 2. Update the user profile
+    // Need to wait 1 second here as that is the minimum time resolution of the 'lastUpdated' field
+    utils.delay(1000);
```
Comment (lboyette-okta) [inline test/it/user-profile-update.js]: This delay won't work without an `await`

File: test/it/user-profile-update.js (hunk 11 above)
Comment (lboyette-okta) [inline test/it/user-profile-update.js]: Hmm. One option is to ensure all of our users and groups have a prefix like `test-`, then we can run one query before any test recordings start to cleanup all test groups and users.

File: test/it/group-get.js [hunk 15]
```diff
@@ -2,9 +2,11 @@ const utils = require('../utils');
 const okta = require('../../');
 const expect = require('chai').expect;
 let orgUrl = process.env.OKTA_CLIENT_ORGURL;
+let mockServer = false;
 
 if (process.env.OKTA_USE_MOCK) {
   orgUrl = `${orgUrl}/group-get-and-stats`;
+  mockServer = true;
 }
```
Comment (bdemers) [inline test/it/group-get.js]: We may want to pull this out into a util method or parent class, it looks like it is reused in each test

File: test/it/group-rule-operations.js [hunk 16]
```diff
@@ -27,15 +29,19 @@ describe('Group-Rule API tests', () => {
...
-
     const newGroup = {
       profile: {
         name: 'Group-Rule API Test Group'
       }
     };
 
+    // Cleanup the user & group if they exist
+    if (!mockServer) {
+      await utils.cleanup(client, newUser, newGroup);
```
Comment (bdemers) [inline test/it/group-rule-operations.js]: We may have multiple groups to remove in the future (possibly same with users)

File: test/it/group-get.js [hunk 17]
```diff
@@ -21,6 +23,11 @@ describe('Group API tests', () => {
       }
     };
 
+    // Cleanup the group if it exists
+    if (!mockServer) {
+      await utils.cleanup(client, null, newGroup);
+    }
+
```
Comment (bdemers) [inline test/it/group-get.js]: Same with the cleanup

File: test/it/user-profile-update.js (hunk 14 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/user-profile-update.js]: Ahhh... Missed it. Adding.

File: test/it/user-profile-update.js (hunk 13 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/user-profile-update.js]: Have updated the test to fetch the user after update to verify that profile was updated.

File: test/it/group-get.js (hunk 15 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/group-get.js]: Yeah, removed it from each test and moved it to a single check in utils.cleanup() method. Much cleaner now.

File: test/it/group-get.js (hunk 17 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/group-get.js]: Removed the dependency on variable

File: test/it/group-rule-operations.js (hunk 16 above)
Comment (vijetmahabaleshwar-okta) [inline test/it/group-rule-operations.js]: Added support to pass an array of users/groups.

File: test/it/user-change-password.js [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    utils.validateUser(createdUser, newUser);
+
+    // 2. Change the user's password
+    const changePasswordCredentials = {
+      oldPassword: { value: 'Abcd1234' },
+      newPassword: { value: '1234Abcd' }
+    };
+
+    let response = await createdUser.changePassword(changePasswordCredentials);
+    expect(response.provider.type).to.equal('OKTA');
```
Comment (bdemers) [inline test/it/user-change-password.js]: We should probably GET the user again, and validate the `passwordChanged` has been updated.  I don't think the value of provider.type is that relevant

**Comment by bdemers** [comment]
I say let's merge it, and deal with any additional changes going forward :shipit:
